### PR TITLE
feat: Change the model build, create separated encoder and decoder

### PR DESCRIPTION
**Pull request by andreanidouglas**
You dont need to keep track of the decoder layers in order to build it
later. Just craete them separatedelly and add to a everything to a
single model.

Model trained with 7000 images can be found on the link below:
https://1drv.ms/u/s!AmmF3hm-u1wQia5nSnLq1HAGk0Gsjw?e=GFO1vH

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 console.log("Hello Autoencoder 🚂");
 
-import * as tf from "@tensorflow/tfjs-node";
+import * as tf from "@tensorflow/tfjs-node-gpu";
```
Comment (shiffman) [inline index.js:3]: Do I need to install CUDA or anything like that to run tf.js with node GPU?

File: index.js [hunk 2]
```diff
@@ -82,87 +74,51 @@ function createDecoder(decoderLayers) {
...
+// encoder
+  const encoder_input = tf.input({shape: [W*W], name: "encoder_input"});
+ // const l0 = tf.layers.flatten().apply(encoder_input);
+  const l1 = tf.layers.dense({units: 128, activation: "relu"}).apply(encoder_input);
+  const l2 = tf.layers.dense({units: 64, activation: "relu"}).apply(l1);
+  const l3 = tf.layers.dense({units: 16, activation: "relu"}).apply(l2);
+  const l4 = tf.layers.dense({units: 4, activation: "relu"}).apply(l3);
+  let encoded = tf.layers.dense({units: 2, activation: "relu", name: "encoder_output"}).apply(l4);
+  
+  let encoder = tf.model({inputs: encoder_input, outputs: encoded, name: "encoder"});
```
Comment (shiffman) [inline index.js:88]: I haven't used `tf.model` before, is it required for this methodology, or can I keep `tf.sequential`? Is `tf.model` preferred?

File: index.js [hunk 3]
```diff
@@ -82,87 +74,51 @@ function createDecoder(decoderLayers) {
...
+ 
+  const learningRate = 0.0001;
+  const optimizer = tf.train.adam(learningRate, 0.000001); // adam(learning_rate, decay)
+  
   autoencoder.compile({
     optimizer: "adam",
     loss: "meanSquaredError",
   });
-  return { decoderLayers, autoencoder };
+  decoder.compile({
```
Comment (shiffman) [inline index.js:117]: I was wondering when I built this if the decoder has to be "compiled" with an optimizer and loss given I'm not training it, only using it for inference?

File: index.js [hunk 4]
```diff
@@ -179,7 +136,7 @@ async function loadImages(total) {
   for (let i = 0; i < total; i++) {
     const num = numeral(i).format("0000");
     const img = await Jimp.read(
-      `AutoEncoder_TrainingData/data/square${num}.png`
+      `data/square${num}.png`
```
Comment (shiffman) [inline index.js:139]: Is there a reason you changed this? I like being able to pull the training data directly from the Processing sketch.

**Review by shiffman — COMMENTED** [review]
Thank you so much for this, it's incredibly helpful! I added some questions and comments but don't feel you have to answer all of them. Like with #3, I'm trying to decide how best to incorporate pull requests like this, if I should merge them or use them as a guide for a future session / video. I like having the code reflect what I did in the streams, but these are critical and important improvements!

File: index.js (hunk 2 above)
Comment (jhrtn) [reply index.js:88]: from [https://js.tensorflow.org/api/latest/#sequential](https://js.tensorflow.org/api/latest/#sequential)

> The key difference between tf.model() and tf.sequential() is that tf.model() is more generic, supporting an arbitrary graph (without cycles) of layers. tf.sequential() is less generic and supports only a linear stack of layers.

File: index.js (hunk 4 above)
Comment (andreanidouglas) [reply index.js:139]: no specific reason, i didn't have processing installed, so i created a Python script to generate the images, i only changed to the output of that

File: index.js (hunk 1 above)
Comment (andreanidouglas) [reply index.js:3]: yes, TensorFlow GPU only supports cuda. 

From my Python experience it needs an extremely specific version of cuda, but for JS it worked with out of the box 11.2 version. I only used the GPU version to speed up training, the CPU version works fine.
